torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -295,7 +272,21 @@ def _call_function_with_auto_output_flattening(
         ),
         example_value=flat_example_value,
     )
+    return flat_variable
+
+
+def overwrite_tensor_vt_requires_grad(graph_output_vts, flat_variable):
+    # this is required for faithfully representing the autograd.Function forward
+    # outputs.
+    for orig_vt, subgraph_vt in zip(graph_output_vts, flat_variable.items):
+        if isinstance(orig_vt, variables.TensorVariable):
+            assert isinstance(subgraph_vt, variables.TensorVariable)
+            orig_vt.requires_grad = subgraph_vt.requires_grad
+            if orig_vt.requires_grad:
+                orig_vt.has_grad_fn = True
 
+
```

Comment:
Can you also add a test for what happens if an autograd.Function returns a non-differentiable tensor (e.g. dtype=torch.int32) as an output, without marking it as non-differentiable? This should always have requires_grad=False